src/xrc/src/api.rs

```diff
@@ -260,19 +320,23 @@ async fn handle_cryptocurrency_pair(
 
     if !utils::is_caller_privileged(&caller) {
         let rate_limited = is_rate_limited(num_rates_needed);
-        let already_inflight = is_inflight(&request.base_asset, timestamp)
-            || is_inflight(&request.quote_asset, timestamp);
-        let charge_cycles_option = if rate_limited || already_inflight {
-            ChargeOption::MinimumFee
-        } else {
-            ChargeOption::OutboundRatesNeeded(num_rates_needed)
-        };
+        let already_inflight = is_inflight(&request.base_asset, timestamp.value)
+            || is_inflight(&request.quote_asset, timestamp.value);
+        let is_past_timestamp_with_rates_needed =
+            timestamp.r#type == NormalizedTimestampType::Past && num_rates_needed > 0;
+        let charge_cycles_option =
+            if rate_limited || already_inflight || is_past_timestamp_with_rates_needed {
+                ChargeOption::MinimumFee
+            } else {
+                ChargeOption::OutboundRatesNeeded(num_rates_needed)
+            };
+
         env.charge_cycles(charge_cycles_option)?;
         if rate_limited {
             return Err(ExchangeRateError::RateLimited);
         }
 
-        if already_inflight {
+        if already_inflight || is_past_timestamp_with_rates_needed {
```

Comment:
How about `is_past_timestamp_not_cached` instead ("cached" implies that no rates are needed)?
It would be clearer here that we return `Pending` if the past time slot cannot be served from the cache.

Reply:
Makes sense. Updated.